RecoTracker/TkSeedGenerator/plugins/DeepCoreSeedGenerator.cc

```diff
@@ -0,0 +1,638 @@
+
+  auto output = std::make_unique<edmNew::DetSetVector<SiPixelCluster>>();
+
+  int jet_number = 0;
+  for (unsigned int ji = 0; ji < cores->size(); ji++) {  //loop jet
+    jet_number++;
+
+    if ((*cores)[ji].pt() > ptMin_) {
+      std::set<long long int> ids;
+      const reco::Candidate& jet = (*cores)[ji];
+      const reco::Vertex& jetVertex = (*vertices)[0];
+
+      std::vector<GlobalVector> splitClustDirSet = splittedClusterDirections(jet, tTopo, pp, jetVertex, 1);
+      bool l2off = (splitClustDirSet.empty());
+      if (splitClustDirSet.empty()) {  //if layer 1 is broken find direcitons on layer 2
+        splitClustDirSet = splittedClusterDirections(jet, tTopo, pp, jetVertex, 2);
+      }
+      splitClustDirSet.push_back(GlobalVector(jet.px(), jet.py(), jet.pz()));
+      for (int cc = 0; cc < (int)splitClustDirSet.size(); cc++) {
+        //-------------------TensorFlow setup - tensor (2/2)----------------------//
+        tensorflow::NamedTensorList input_tensors;
+        input_tensors.resize(3);
+        input_tensors[0] =
+            tensorflow::NamedTensor(inputTensorName_[0], tensorflow::Tensor(tensorflow::DT_FLOAT, input_size_eta));
+        input_tensors[1] =
+            tensorflow::NamedTensor(inputTensorName_[1], tensorflow::Tensor(tensorflow::DT_FLOAT, input_size_pt));
+        input_tensors[2] = tensorflow::NamedTensor(inputTensorName_[2],
+                                                   tensorflow::Tensor(tensorflow::DT_FLOAT, {input_size_cluster}));
+
+        //put all the input tensor to 0
+        input_tensors[0].second.matrix<float>()(0, 0) = 0.0;
+        input_tensors[1].second.matrix<float>()(0, 0) = 0.0;
+        for (int x = 0; x < jetDimX; x++) {
+          for (int y = 0; y < jetDimY; y++) {
+            for (int l = 0; l < 4; l++) {
+              input_tensors[2].second.tensor<float, 4>()(0, x, y, l) = 0.0;
+            }
+          }
+        }
+        //-----------------end of TF setup (2/2)----------------------//
+
+        GlobalVector bigClustDir = splitClustDirSet.at(cc);
+
+        LocalPoint jetInter(0, 0, 0);
+
+        jet_eta = jet.eta();
+        jet_pt = jet.pt();
+        input_tensors[0].second.matrix<float>()(0, 0) = jet.eta();
+        input_tensors[1].second.matrix<float>()(0, 0) = jet.pt();
+
+        edmNew::DetSetVector<SiPixelCluster>::const_iterator detIt = inputPixelClusters->begin();
+
+        const GeomDet* globDet =
+            DetectorSelector(2, jet, bigClustDir, jetVertex, tTopo);  //select detector mostly hitten by the jet
+
+        if (globDet == nullptr)
+          continue;
+
+        const GeomDet* goodDet1 = DetectorSelector(1, jet, bigClustDir, jetVertex, tTopo);
+        const GeomDet* goodDet3 = DetectorSelector(3, jet, bigClustDir, jetVertex, tTopo);
+        const GeomDet* goodDet4 = DetectorSelector(4, jet, bigClustDir, jetVertex, tTopo);
+
+        for (; detIt != inputPixelClusters->end(); detIt++) {  //loop deset
```

Comment:
@kpedro88  so why here the product() is not needed? I'm confused...

Reply:
I think it should be used here also.

Reply:
The call to `product()`? Why?

Reply:
In the context of making a range-based loop.

Reply:
Range-for can be done with `for (auto const& elem : *handle)`

Reply:
sorry people but here neither 
`for (const auto& detset : *inputPixelClusters.product())`
neither
`for (const auto& detset : *inputPixelClusters)`
works here (erroror[1]) at comilation time. Instead the @jpata solution works fine **but** the same solution breaks at run time inside the functions. with the error [2]


```
[1]  error: no match for 'operator*' (operand type is 'const edmNew::DetSetVector<SiPixelCluster>')
         for (const auto& detset : *inputPixelClusters){
[2] Exception Message:
An attempt was made to dereference an edm::Handle which was never set.
```

Reply:
In [1] `inputPixelClusters`  is not a Handle so there is no need to de-reference a handle, and `for (const auto& detset : inputPixelClusters)` should do the job

In [2] the member functions use `inputPixelClusters` member variable (not the local variable in from `produce()`) which is a non-initialized Handle. Preferred way would to be pass `inputPixelClusters` to member functions by function arguments (alternatively it could be passed from member variable which would be a raw pointer).

Reply:
Yes, I thought about passing it to the function as an argument... and now seems to me the best solution, because if i use the member variable have I to re-initzialize the Handle inside the function?

Reply:
ok